transformer_engine/pytorch/cpp_extensions/gemm.py

```diff
@@ -25,6 +27,143 @@ def _empty_tensor() -> torch.Tensor:
     return torch.Tensor()
 
 
+def general_gemm(
+    A: Union[torch.Tensor, Float8Tensor],
+    B: Union[torch.Tensor, Float8Tensor],
+    workspace: torch.Tensor,
+    gelu: bool = False,
+    accumulate: bool = False,
+    out: Optional[torch.Tensor] = None,
+    bias: Optional[torch.Tensor] = None,
+    use_split_accumulator: bool = False,
+    D_dtype: Optional[tex.DType] = None,
+    ub_algo: tex.UbufOverlapAlgo = None,
+    ub: Union[tex.UbufCommOverlap, tex.UbufP2PCommOverlap] = None,
+    extra_output_tensor: Optional[torch.Tensor] = None,
+) -> torch.Tensor:
```

Comment:
While we're reworking this API, perhaps we should call it `matmul` since it's less ambiguous (e.g. with column-major/row-major order). We should also keep the core API simple like [`torch.matmul`](https://pytorch.org/docs/stable/generated/torch.matmul.html) and [`np.matmul`](https://numpy.org/doc/2.0/reference/generated/numpy.matmul.html), and leave our non-standard options as kwargs:
```python
def matmul(
    A: torch.Tensor,  # maybe QuantizedTensor
    B: torch.Tensor,  # maybe QuantizedTensor
    /,
    out: Optional[torch.Tensor] = None,  # maybe QuantizedTensor
    *,
    transa: bool = False,
    transb: bool = False,
    out_dtype: Optional[tex.DType] = None,
    accumulate_out: bool = False,  # alternatively: alpha and beta
    bias: Optional[torch.Tensor] = None,
    activation: Optional[str] = None,  # more general than gelu
    workspace: torch.Tensor,  # consider allocating in C++
    use_split_accumulator: bool = False,  # maybe hide within cublas_options kwarg?
    userbuffers_options: Optional[dict] = None,  # minimize impact of unstable UB API
) -> torch.Tensor:  # maybe QuantizedTensor
```